### PR TITLE
Adds cypress reporter

### PR DESCRIPTION
**Pull request by jesseduffield**
Adds a reporter for cypress. Cypress uses mocha internally so there's lots of overlap with the mocha reporter, except for how it parses errors. Happy to try and DRY that up if requested.

### CODE REVIEW

File: cypress/reporter.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+  testRunFinished() {
+    uploadTestResults(this._testEnv, this._testResults, this._options)
+  }
+
+  analyticsResult(state) {
+    // Mocha test statuses:
+    // - passed
+    // - failed
+    // - pending
```
[the remark this answers is not in the document]
Comment (jesseduffield) [reply cypress/reporter.js:98]: I'm not sure, this was just a result of me copy+pasting from the existing mocha reporter.

**Comment by CallumHoward** [comment]
I have tested this with a suite of ~100 test cases. It appears to work well, would love to see this merged. I see there is a failing PR check currently however.

**Comment by jesseduffield** [comment]
@CallumHoward I've got those tests passing.
Cypress has some OS-level depdendencies it requires so I've switched to using the cypress base image which itself contains node.

**Review by CallumHoward — APPROVED** [review]
Looks great, works great. Thanks for this!

**Comment by swebb** [comment]
Hi @jesseduffield and @CallumHoward. Sorry for the slow response.

Thank you for the contribution. I'll raise a ticket internally and see if we can get this reviewed & merged.

**Comment by jesseduffield** [comment]
Thanks @swebb :)

**Comment by jesseduffield** [comment]
hey @swebb do you know roughly when this PR will be reviewed?

**Comment by blaknite** [comment]
Hey @jesseduffield big thanks for the work you've put into this PR 😊

Unfortunately the Test Analytics team does not have capacity to review, merge and support this feature in Q1 given our current commitments and workload. We're going to consider it as part of our Q2 priorities.

We'll keep you in the loop!

**Comment by jesseduffield** [comment]
@blaknite all good, thanks for letting me know :)

**Comment by mitchwd** [comment]
@blaknite Any movement on this one? This would be a huge help to us!

File: README.md [hunk 2]
```diff
@@ -115,6 +115,28 @@ Official [Buildkite Test Analytics](https://buildkite.com/test-analytics) collec
...
 
+   ### Cypress
+
+   Update your [Cypress configuration](https://docs.cypress.io/guides/references/configuration):<br>
+
+   ```js
+    // cypress.config.js
+
+    // Send results to Test Analytics
+   reporter: "test-collector-javascript/cypress/reporter,
```
Comment (nprizal) [inline README.md:126]:
```suggestion
   reporter: "buildkite-test-collector/cypress/reporter",
```

File: README.md [hunk 3]
```diff
@@ -115,6 +115,28 @@ Official [Buildkite Test Analytics](https://buildkite.com/test-analytics) collec
...
+   ```
+
+   If you would like to pass in the API token using a custom environment variable, you can do so using the reporterOptions.
+
+   ```js
+   // cypress.config.js
+
+   // Send results to Test Analytics
+   reporterOptions: {
+    token_name: "CUSTOM_ENV_VAR_NAME"
```
Comment (nprizal) [inline README.md:136]: What do you think of passing the `token` directly to the option instead of env name, similar to what `jest/reporter` does?
```suggestion
    token_name: process.env.BUILDKITE_ANALYTICS_TOKEN
```

File: examples/cypress/cypress/support/commands.js
Comment (nprizal) [inline examples/cypress/cypress/support/commands.js:1]: Is this file needed, or can be removed?

File: examples/cypress/cypress.config.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const { defineConfig } = require("cypress");
+
+module.exports = defineConfig({
+  reporter: "../../cypress/reporter",
+  reporterOptions: {
```
Comment (nprizal) [inline examples/cypress/cypress.config.js:5]: Maybe this options can be removed? The specs expect that the token can be set via `reporterOptions` and env variable. If we set the token here, it will defeat the purpose of those specs.

**Review by nprizal — COMMENTED** [review]
@jesseduffield Thanks for your contribution. I tested the integration with Test Analytics, and everything seems to be working fine
Screenshot 2023-09-29 at 11 58 13 AM

I have some small feedbacks that you can consider.

File: examples/cypress/cypress.config.js (hunk 4 above)
Comment (jesseduffield) [reply examples/cypress/cypress.config.js:5]: `reporterOptions` isn't setting the token's value, it's specifying the token's name. So I think this is fine: it's the same thing that the mocha reporter does.

File: examples/cypress/cypress/support/commands.js
Comment (jesseduffield) [reply examples/cypress/cypress/support/commands.js:1]: Done

File: README.md (hunk 2 above)
Comment (jesseduffield) [reply README.md:126]: Done

File: buildkite.yaml [hunk 5]
```diff
@@ -6,8 +6,8 @@ steps:
       - "npm test"
     plugins:
       - docker#v3.7.0:
-          image: node:lts
+          image: cypress/base:16.13.0
```
Comment (nprizal) [inline buildkite.yaml:9]: Hi @jesseduffield unfortunately this breaks the CI pipeline
https://buildkite.com/buildkite/javascript-test-collectors/builds/203#018b2121-524b-463a-9e5f-c5e4f63e6510.
Is there any docker image that we can use to avoid this? Or maybe we can create a separate step for Cypress, so it doesn't affect the existing step?

Comment (jesseduffield) [reply buildkite.yaml]: I've switched back to node:lts and added a step for adding cypress dependencies

**Comment by jesseduffield** [comment]
This is ready for another CI run @nprizal @blaknite

**Comment by jesseduffield** [comment]
Hey guys, I no longer have the time to work on this so I'm going to close it, but anybody else who wants to pick it up, feel free!
